### PR TITLE
Add security hub summary API

### PR DESCRIPTION
**Pull request by stonezdj**
include with_dangerous_artifact, with_dangerous_cve option

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [x] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: api/v2.0/swagger.yaml [hunk 1]
```diff
@@ -6052,6 +6052,104 @@ paths:
...
+      summary: Get vulnerability system summary
+      description: Retrieve the vulnerability summary of the system
+      tags:
+        - securityhub
+      operationId: getSecuritySummary
+      parameters:
+        - $ref: '#/parameters/requestId'
+        - name: with_dangerous_cve
+          in: query
+          description: Specify whether the scan overview is inclued inside the returning artifacts
```
Comment (chlins) [inline api/v2.0/swagger.yaml:6066]: Mismatched description.

File: api/v2.0/swagger.yaml [hunk 2]
```diff
@@ -6052,6 +6052,104 @@ paths:
...
+          type: boolean
+          required: false
+          default: false
+        - name: with_dangerous_artifact
+          in: query
+          description: Specify whether the scan overview is inclued inside the returning artifacts
```
Comment (chlins) [inline api/v2.0/swagger.yaml:6072]: Ditto.

File: api/v2.0/swagger.yaml [hunk 3]
```diff
@@ -6052,6 +6052,104 @@ paths:
...
+      responses:
+        '200':
+          description: Success
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/CVERecord'
+          headers:
+            X-Total-Count:
+              description: The total count of projects
```
Comment (chlins) [inline api/v2.0/swagger.yaml:6111]: Ditto.

File: api/v2.0/swagger.yaml [hunk 4]
```diff
@@ -9600,3 +9698,176 @@ definitions:
...
+        description: the count of critical vulnerabilities
+      high_cnt:
+        type: integer
+        format: int64
+        x-omitempty: false
+        description: the count of high vulnerabilities
+      medium_cnt:
+        type: integer
+        x-omitempty: false
+        description: the count of high vulnerabilities
```
Comment (chlins) [inline api/v2.0/swagger.yaml:9813]: Ditto.

Comment (stonezdj) [reply api/v2.0/swagger.yaml:9813]: Updated

File: api/v2.0/swagger.yaml (hunk 3 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml:6111]: Updated

File: api/v2.0/swagger.yaml (hunk 2 above)
Comment (stonezdj) [reply api/v2.0/swagger.yaml:6072]: Updated

File: src/controller/securityhub/controller.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	secHubModel "github.com/goharbor/harbor/src/pkg/securityhub/model"
+)
+
+// Ctl is the global controller for security hub
+var Ctl = NewController()
+
+// Controller controller of security hub
+type Controller interface {
+	// SecuritySummary returns the security summary of the specified project.
+	SecuritySummary(ctx context.Context, projectID int64, withCVE bool, withArtifact bool) (*secHubModel.Summary, error)
```
Comment (chlins) [inline src/controller/securityhub/controller.go:33]: If the method may have more extended capabilities, we can wrap the `withCVE` and `withArtifact` to `Options`, but currently it's ok to me and can update in the future.

File: src/controller/securityhub/controller.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+type controller struct {
+	artifactMgr artifact.Manager
+	scannerMgr  scanner.Manager
+	secHubMgr   securityhub.Manager
+}
+
+// NewController ...
+func NewController() Controller {
+	return &controller{
+		artifactMgr: artifact.NewManager(),
```
Comment (chlins) [inline src/controller/securityhub/controller.go:45]: The global artifact manager is `src/pkg.ArtifactMgr`.

File: src/pkg/securityhub/dao/security.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+           from artifact a 
+      left join scan_report s on a.digest = s.digest 
+          where s.registration_uuid= ? and s.uuid is not null`
+
+	dangerousCVESQL = `select vr.*
+from vulnerability_record vr
+where vr.cvss_score_v3 is not null
+and vr.registration_uuid = ?
+order by vr.cvss_score_v3 desc
+limit 5`
```
Comment (chlins) [inline src/pkg/securityhub/dao/security.go:56]: We can add some comments here to explain background and concerns why limit it to 5.

File: src/server/v2.0/handler/security.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	controller securityhub.Controller
+}
+
+func (s *securityAPI) GetSecuritySummary(ctx context.Context,
+	params securityModel.GetSecuritySummaryParams) middleware.Responder {
+	secCtx, ok := security.FromContext(ctx)
+	if !ok {
+		return s.SendError(ctx, errors.UnauthorizedError(errors.New("security context not found")))
+	}
+	if !secCtx.IsAuthenticated() && !secCtx.IsSysAdmin() {
```
Comment (chlins) [inline src/server/v2.0/handler/security.go:49]: I think the logic may have some potential issue if we only want open this API for system admin.

**Review by chlins — APPROVED** [review]
lgtm

File: src/controller/securityhub/controller.go (hunk 5 above)
Comment (wy65701436) [reply src/controller/securityhub/controller.go:33]: use option as the parameter?

File: src/pkg/securityhub/dao/security.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+       sum(s.medium_cnt)   medium_cnt,
+       sum(s.low_cnt)      low_cnt,
+       sum(s.none_cnt)     none_cnt,
+       sum(s.unknown_cnt)  unknown_cnt,
+       sum(s.fixable_cnt)  fixable_cnt
+from artifact a
+         left join scan_report s on a.digest = s.digest
+         where s.registration_uuid = ?`
+
+	dangerousArtifactSQL = `select a.project_id project, a.repository_name repository, a.digest, s.critical_cnt, s.high_cnt, s.medium_cnt, s.low_cnt
```
Comment (wy65701436) [inline src/pkg/securityhub/dao/security.go:38]: can we just order by critical_cnt?

File: src/server/v2.0/handler/security.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+func (s *securityAPI) GetSecuritySummary(ctx context.Context,
+	params securityModel.GetSecuritySummaryParams) middleware.Responder {
+	secCtx, ok := security.FromContext(ctx)
+	if !ok {
+		return s.SendError(ctx, errors.UnauthorizedError(errors.New("security context not found")))
+	}
+	if !secCtx.IsSysAdmin() {
+		return s.SendError(ctx, errors.UnauthorizedError(nil).WithMessage(secCtx.GetUsername()))
```
Comment (wy65701436) [inline src/server/v2.0/handler/security.go:50]: should be 403?

File: src/server/v2.0/handler/security.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+	controller securityhub.Controller
+}
+
+func (s *securityAPI) GetSecuritySummary(ctx context.Context,
+	params securityModel.GetSecuritySummaryParams) middleware.Responder {
+	secCtx, ok := security.FromContext(ctx)
+	if !ok {
+		return s.SendError(ctx, errors.UnauthorizedError(errors.New("security context not found")))
+	}
+	if !secCtx.IsSysAdmin() {
```
Comment (wy65701436) [inline src/server/v2.0/handler/security.go:49]: please use the RequireSystemAccess()

File: api/v2.0/swagger.yaml [hunk 12]
```diff
@@ -9600,3 +9633,116 @@ definitions:
...
+      critical_cnt:
+        type: integer
+        x-omitempty: false
+        description: the count of critical vulnerabilities
+      high_cnt:
+        type: integer
+        format: int64
+        x-omitempty: false
+        description: the count of high vulnerabilities
+      medium_cnt:
```
Comment (wy65701436) [inline api/v2.0/swagger.yaml:9745]: do we consider to add the low_cnt as well?

File: src/pkg/securityhub/dao/security.go (hunk 9 above)
Comment (stonezdj) [reply src/pkg/securityhub/dao/security.go:38]: If the critical_cnt is the same, sort the high_cnt

File: src/server/v2.0/handler/security.go (hunk 11 above)
Comment (stonezdj) [reply src/server/v2.0/handler/security.go:49]: Updated

File: src/controller/securityhub/controller.go (hunk 5 above)
Comment (stonezdj) [reply src/controller/securityhub/controller.go:33]: Updated

File: src/pkg/securityhub/model/model.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+	FixableCnt         int64                       `json:"fixable_cnt"`
+	ScannedCnt         int64                       `json:"scanned_cnt"`
+	NotScanCnt         int64                       `json:"not_scan_cnt"`
+	TotalArtifactCnt   int64                       `json:"total_artifact_cnt"`
+	DangerousCVEs      []*scan.VulnerabilityRecord `json:"dangerous_cves"`
+	DangerousArtifacts []*DangerousArtifact        `json:"dangerous_artifacts"`
+}
+
+// DangerousArtifact define the most dangerous artifact
+type DangerousArtifact struct {
```
Comment (wy65701436) [inline src/pkg/securityhub/model/model.go:36]: why not just wrapper the artifact, and add the count infor?

**Review by wy65701436 — APPROVED** [review]
lgtm

File: src/pkg/securityhub/model/model.go (hunk 13 above)
Comment (stonezdj) [reply src/pkg/securityhub/model/model.go:36]: We have to define orm mapping tag, it is different with the original one.

**Pull request merged**